File: ocaml/ch4/bigstep.ml
type term =
    TmTrue
  | TmFalse
  | TmIf of term  * term * term
  | TmZero
  | TmSucc of term
  | TmPred of term
  | TmIsZero of term

let rec isnumericval t = match t with
    TmZero -> true
  | TmSucc(t1) -> isnumericval t1
  | _ -> false

let isval t = match t with
    TmTrue -> true
  | TmFalse -> true
  | t when isnumericval t -> true
  | _ -> false

exception NoRuleApplies

let rec eval t = match t with
    TmTrue -> TmTrue
  | TmFalse -> TmFalse
  | TmZero -> TmZero
  | TmIf(t1, t2, t3) ->
      let t1' = eval t1 in
      (match t1' with
          TmTrue -> eval t2
        | TmFalse -> eval t3
        | _ -> raise NoRuleApplies)
  | TmSucc(t1) when isnumericval t1 -> TmSucc(eval t1)
  | TmPred(t1) ->
      let t1' = eval t1 in
      (match t1' with
          TmZero -> TmZero
        | TmSucc(t2) when isnumericval t2 -> t2
        | _ -> raise NoRuleApplies)
  | TmIsZero(t1) ->
      let t1' = eval t1 in
      (match t1' with
          TmZero -> TmTrue
        | TmSucc(t2) when (isnumericval t2) -> TmFalse
        | _ -> raise NoRuleApplies)
  | _ -> raise NoRuleApplies

let rec to_string t = match t with
    TmTrue -> "True"
  | TmFalse -> "False"
  | TmIf(t1, t2, t3) ->
    "If " ^ to_string t1 ^ " then " ^ to_string t2 ^ " else " ^ to_string t3
  | TmZero -> "0"
  | TmSucc(t1) ->
    "Succ(" ^ to_string t1 ^ ")"
  | TmPred(t1) ->
    "Pred(" ^ to_string t1 ^ ")"
  | TmIsZero(t1) ->
    "IsZero(" ^ to_string t1 ^ ")"

let main() =
    let t = TmIf(TmIsZero(TmZero), TmSucc(TmZero), TmTrue) in
    let r = eval t in
    let s = to_string r in
    Printf.printf "%s\n" s;
    exit 0;;
main();;

File: rust/ch4/bigstep.rs
#![allow(dead_code)]

#[derive(Clone, Debug)]
enum Term {
    TmTrue,
    TmFalse,
    TmIf {
        cond: Box<Term>,
        cons: Box<Term>,
        alt: Box<Term>
    },
    TmZero,
    TmSucc {
        val: Box<Term>
    },
    TmPred {
        val: Box<Term>
    },
    TmIsZero {
        val: Box<Term>
    }
}

use Term::*;

fn isnumerical(t: &Term) -> bool {
    match t {
        TmZero => true,
        TmSucc { val } => isnumerical(val),
        _ => false,
    }
}

fn isval(t: &Term) -> bool {
    match t {
        TmTrue => true,
        TmFalse => true,
        t if isnumerical(t) => true,
        _ => false
    }
}

fn eval(t: &Term) -> Term {
    match t {
        t if isval(t) => t.clone(),
        TmIf { cond, cons, alt } => {
            match cond.as_ref() {
                TmTrue => eval(cons),
                TmFalse => eval(alt),
                _ => panic!(),
            }
        }
        TmSucc { val } => {
            if isnumerical(&eval(&val)) {
                return eval(val)
            } else {
                panic!()
            }
        }
        TmPred { val } => {
            let val1 = eval(val);
            match val1 {
                TmZero => val1,
                TmSucc { val } if isnumerical(val.as_ref()) => *val,
                _ => panic!(),
            }
        }
        TmIsZero { val } => {
            let val1 = eval(val);
            match val1 {
                TmZero => TmTrue,
                TmSucc { val } if isnumerical(val.as_ref()) => TmFalse,
                _ => panic!(),
            }
        }
        _ => panic!()
    }
}

fn main() {
    let t = TmIf {
        cond: Box::new(TmTrue),
        cons: Box::new(TmSucc {
            val: Box::new(TmZero)
        }),
        alt: Box::new(TmFalse)
    };
    let t1 = eval(&t);
    println!("{:?}", t1);
}

